File: solutions/typescript/0078.ts
/**
 * 78. 子集
 * 给定一个不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
 * 解集不能包含重复子集，可以按任意顺序返回。
 */
function subsets(nums: number[]): number[][] {
    const result: number[][] = [];
    const path: number[] = [];

    const dfs = (start: number): void => {
        result.push([...path]);
        for (let i = start; i < nums.length; i++) {
            path.push(nums[i]);
            dfs(i + 1);
            path.pop();
        }
    };

    dfs(0);
    return result;
}

export default subsets;

File: solutions/rust/0078.rs
/// 78. 子集
/// 回溯枚举所有子集（幂集），数组元素互不相同
pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut result: Vec<Vec<i32>> = Vec::new();
    let mut path: Vec<i32> = Vec::new();

    fn dfs(start: usize, nums: &[i32], path: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
        result.push(path.clone());
        for i in start..nums.len() {
            path.push(nums[i]);
            dfs(i + 1, nums, path, result);
            path.pop();
        }
    }

    dfs(0, &nums, &mut path, &mut result);
    result
}

impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        subsets(nums)
    }
}